substrate/frame/revive/src/vm/evm/instructions/host.rs

```diff
@@ -15,124 +15,79 @@
-
-	let Some(balance) = context.host.balance(context.interpreter.input.target_address()) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	push!(context.interpreter, balance.data);
+	gas!(context.interpreter, RuntimeCosts::Balance);
+	let balance = context.interpreter.extend.balance();
+	push!(context.interpreter, balance.into_revm_u256());
 }
 
 /// Implements the EXTCODESIZE instruction.
 ///
 /// Gets the size of an account's code.
 pub fn extcodesize<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	popn_top!([], top, context.interpreter);
-	let address = top.into_address();
-	let Some(code) = context.host.load_account_code(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	let spec_id = context.interpreter.runtime_flag.spec_id();
-	if spec_id.is_enabled_in(BERLIN) {
-		gas_legacy!(context.interpreter, warm_cold_cost(code.is_cold));
-	} else if spec_id.is_enabled_in(TANGERINE) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 20);
-	}
-
-	*top = U256::from(code.len());
+	gas!(context.interpreter, RuntimeCosts::CodeSize);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_size = context.interpreter.extend.code_size(&h160);
+	*top = U256::from(code_size);
 }
 
 /// EIP-1052: EXTCODEHASH opcode
 pub fn extcodehash<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	check!(context.interpreter, CONSTANTINOPLE);
 	popn_top!([], top, context.interpreter);
-	let address = top.into_address();
-	let Some(code_hash) = context.host.load_account_code_hash(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	let spec_id = context.interpreter.runtime_flag.spec_id();
-	if spec_id.is_enabled_in(BERLIN) {
-		gas_legacy!(context.interpreter, warm_cold_cost(code_hash.is_cold));
-	} else if spec_id.is_enabled_in(ISTANBUL) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 400);
-	}
-	*top = code_hash.into_u256();
+	gas!(context.interpreter, RuntimeCosts::CodeHash);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_hash = context.interpreter.extend.code_hash(&h160);
+	*top = U256::from_be_bytes(code_hash.0);
 }
 
 /// Implements the EXTCODECOPY instruction.
 ///
 /// Copies a portion of an account's code to memory.
 pub fn extcodecopy<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	popn!([address, memory_offset, code_offset, len_u256], context.interpreter);
-	let address = address.into_address();
-	let Some(code) = context.host.load_account_code(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
+	let address = sp_core::H160::from_slice(&address.to_be_bytes::<32>()[12..]);
+	let code_hash = context.interpreter.extend.code_hash(&address);
+	let code = crate::PristineCode::<E::T>::get(&code_hash).unwrap_or_default();
```

Comment:
There shouldn't be any logic inside the instruction implementation. All of that needs to live in `Ext` so it can be shared with the PVM backend. The instruction impl should strictly be concerned with gas charging and i/o.

Reply:
Good catch I agree with this. We want to preserve as much semantics as possible over both VM backends.